src/main/java/com/mykaarma/urlshortener/service/UrlService.java

```diff
@@ -0,0 +1,250 @@
+			
+			log.info(" Short Url for shortUrlHash={} is expired ", shortUrlHash);
+			throw new ShortUrlNotFoundException(UrlErrorCodes.SHORT_URL_NOT_FOUND, "Long URL not found for the given short URL");
+
+		}
+		
+		log.info(String.format("shortUrl for shortUrlHash=%s redirects to longUrl=%s", shortUrlHash, existingUrlDetails.getLongUrl()));
+
+		return replaceUrlInRedirectingHtmlTemplate(existingUrlDetails.getLongUrl());
+	}
+	
+	public UrlDetails getShortUrlDetails(String shortUrlHash) throws ShortUrlException {
+		
+		UrlDetails existingUrlDetails = shortUrlCacheAdapter.fetchUrlDetailsFromCache(shortUrlHash);
+		
+		if(existingUrlDetails != null) {
+			log.info("Fetching existing short url details from cache for shortUrlHash="+shortUrlHash);
+			return existingUrlDetails;
+		}
+		
+		existingUrlDetails = getExistingShortUrlDetails(shortUrlHash);
+
+		if (existingUrlDetails==null) {
+			
+			log.info(String.format("shortUrlHash=%s is not present in DB ", shortUrlHash));
+			throw new ShortUrlNotFoundException(UrlErrorCodes.SHORT_URL_NOT_FOUND, "Long URL not found for the given short URL");
+
+		}
+
+		if(existingUrlDetails.getExpiryDateTime().before(new Date())) {
+			
+			log.info(String.format("Short Url for shortUrlHash=%s is expired ", shortUrlHash));
+			throw new ShortUrlNotFoundException(UrlErrorCodes.SHORT_URL_NOT_FOUND, "Long URL not found for the given short URL");
+
+		}
+		
+		long ttl = urlServiceUtil.getDurationBetweenTwoDatesInSeconds(existingUrlDetails.getExpiryDateTime(), new Date());
+		if(ttl > 24*60*60) {
+			ttl = 24*60*60;
+		}
+		
+		shortUrlCacheAdapter.saveInCache(shortUrlHash, existingUrlDetails, ttl);
+		
+		return existingUrlDetails;
+	}
+	
+	/**
+	 * Returns the HTML response redirecting to the longUrl
+	 * @param longUrl
+	 * @return HTML response redirecting to longUrl
+	 */
+	public String replaceUrlInRedirectingHtmlTemplate(String longUrl) {
+		String html = "<!DOCTYPE html>\r\n<html>\r\n<head lang=\"en\">\r\n    <meta charset=\"UTF-8\">\r\n"
+				+ "<meta http-equiv=\"cache-control\" content=\"max-age=0\" />\r\n"
+				+ "<meta http-equiv=\"cache-control\" content=\"no-cache\" />\r\n"
+				+ "<meta http-equiv=\"expires\" content=\"0\" />\r\n"
+				+ "<meta http-equiv=\"expires\" content=\"Tue, 01 Jan 1980 1:00:00 GMT\" />\r\n"
+				+ "<meta http-equiv=\"pragma\" content=\"no-cache\" />\r\n"
+				+ "<meta http-equiv=\"refresh\" content=\"0; URL='_url'\" />\r\n    <title></title>\r\n</head>\r\n<body>\r\n_loading  Please Wait\r\n</body>\r\n</html>";
+		html = html.replace("_url", longUrl);
+
+		return html;
+	}
+	
+	/**
+	 * Fetches the urlDetails from the database using the shortUrlHash
+	 * 
+	 * @param shortUrlHash
+	 * @return urlDetails
+	 * @throws ShortUrlException
+	 */
+	private UrlDetails getExistingShortUrlDetails(String shortUrlHash) throws ShortUrlException {
+
+		List<UrlDetails> existingShortUrls = urlRepository.getUrlDetailsByShortUrlHash(shortUrlHash);
```

Comment:
Why does this method return a list? There can at most be on shortUrl for a given hash